### PR TITLE
Anisotropic diffusion for FV 

### LINKED ISSUE #21832: Simple framework kernel for anisotropic heat conduction
## Reason

This is supported using `PINSFVEnergyAnisotropicDiffusion` but a simple kernel in the framework
would be nice

## Design

Add a new kernel that takes a vector diffusion coefficient (diagonal tensor).

## Impact

None expected

### PR DESCRIPTION
**Pull request by snschune**
closes #21832 

## Reason

Anisotropic diffusion should be present in the framework for standard FV variables and not
only as special case in the Navier-Stokes module.

## Design

Added an additional kernel. Changed interpolation of diffusion coefficient on the face.

## Impact

none, new capability

### CODE REVIEW

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  }
+}
+
+ADReal
+FVAnisotropicDiffusion::computeQpResidual()
+{
+  ADReal r = 0;
+  const auto face_elem = elemFromFace();
+  const auto face_neighbor = neighborFromFace();
+  const auto & grad_T = _var.adGradSln(*_face_info);
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:70]:
```suggestion
  const auto & grad_T = _var.adGradSln(*_face_info, _var.faceInterpolationMethod() == Moose::FV::InterpMethod::SkewCorrectedAverage);
```

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  for (std::size_t i = 0; i < LIBMESH_DIM; i++)
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:71]:
```suggestion
  for (const auto i : make_range(Moose::dim))
```

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  {
+    ADReal k_face_inv;
+    Moose::FV::interpolate(Moose::FV::InterpMethod::Average,
+                           k_face_inv,
+                           1.0 / _coeff(face_elem)(i),
+                           1.0 / _coeff(face_neighbor)(i),
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:77]: I would pre-evaluate both of these vectors and store them in local variables pre-loop.

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  const auto face_elem = elemFromFace();
+  const auto face_neighbor = neighborFromFace();
+  const auto & grad_T = _var.adGradSln(
+      *_face_info, _var.faceInterpolationMethod() == Moose::FV::InterpMethod::SkewCorrectedAverage);
+  ADRealVectorValue k_face_inv, k_elem_inv, k_neigh_inv;
+  for (const auto i : make_range(Moose::dim))
+  {
+    k_elem_inv(i) = 1.0 / _coeff(face_elem)(i);
+    k_neigh_inv(i) = 1.0 / _coeff(face_neighbor)(i);
+  }
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:77]:
```suggestion
  ADRealVectorValue k_face_inv, k_elem_inv, k_neigh_inv;
  const auto k_elem = _coeff(face_elem), k_neighbor = _coeff(face_neighbor);
  for (const auto i : make_range(Moose::dim))
  {
    k_elem_inv(i) = 1.0 / k_elem(i);
    k_neigh_inv(i) = 1.0 / k_neighbor(i);
  }
```

File: test/tests/fvkernels/fv_anisotropic_diffusion/tests [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[Tests]
+  [fv_anisotropic_diffusion]
+    type = 'Exodiff'
+    input = 'fv_anisotropic_diffusion.i'
+    exodiff = 'fv_anisotropic_diffusion_out.e'
+    requirement = 'The system shall be able to solve anisotropic diffusion problems using the FV method'
+    design = 'FVAnisotropicDiffusion.md'
+    issues = '#21832'
+  [../]
```
Comment (lindsayad) [inline test/tests/fvkernels/fv_anisotropic_diffusion/tests:9]:
```suggestion
    issues = '#21832'
    ad_indexing_type = 'global'
  []
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by snschune** [comment]
Both loops over `LIBMESH_DIM` and `Moose::dim` iterate from 0 to 2. The problem is only 2D, so why does that happen?
`_mesh.dimension()` does the right thing.

**Comment by lindsayad** [comment]
Moose::dim is just a nice wrapper of LIBMESH_DIM, which is a compile time constant which is always 3 for MOOSE. The mesh dimension, meanwhile, is run-time based on the user mesh.

**Review by lindsayad — COMMENTED** [review]
@grmnptr can you finish reviewing/merging this PR for me plz?

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    mooseAssert(k_elem(i) != 0, "Diffusion coefficient is equal to zero");
+    mooseAssert(k_neighbor(i) != 0, "Diffusion coefficient is equal to zero");
+    k_elem_inv(i) = 1.0 / k_elem(i);
+    k_neigh_inv(i) = 1.0 / k_neighbor(i);
+  }
+
+  Moose::FV::interpolate(
+      Moose::FV::InterpMethod::Average, k_face_inv, k_elem_inv, k_neigh_inv, *_face_info, true);
+  for (const auto i : make_range(Moose::dim))
+    r += _normal(i) * grad_T(i) / k_face_inv(i);
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:85]:
```suggestion
  const auto k_elem = _coeff(face_elem), k_neighbor = _coeff(face_neighbor);
  for (const auto i : _mesh.dimension())
  {
    mooseAssert(k_elem(i) != 0, "Diffusion coefficient is equal to zero");
    mooseAssert(k_neighbor(i) != 0, "Diffusion coefficient is equal to zero");
    auto k_elem_inv = 1.0 / k_elem(i);
    auto k_neigh_inv = 1.0 / k_neighbor(i);
    decltype(k_elem_inv) k_face_inv;

    Moose::FV::interpolate(
        Moose::FV::InterpMethod::Average, k_face_inv, k_elem_inv, k_neigh_inv, *_face_info, true);
    r += _normal(i) * grad_T(i) / k_face_inv;
```

Comment (lindsayad) [reply framework/src/fvkernels/FVAnisotropicDiffusion.C:85]: This will most likely need a clang-formatting

**Review by lindsayad — CHANGES_REQUESTED** [review]
I think this would fix your divide by zero ... ?

File: framework/src/fvkernels/FVAnisotropicDiffusion.C (hunk 6 above)
Comment (lindsayad) [reply framework/src/fvkernels/FVAnisotropicDiffusion.C:85]: Aw looks like you've already got this passing the debug test

**Review by grmnptr — COMMENTED** [review]
Thanks for the contribution! I have one question.

File: framework/doc/content/source/fvkernels/FVAnisotropicDiffusion.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+!equation
+- \nabla \cdot \mathbf{k} \nabla \phi,
+
+where $\mathbf{k}$ is the diagonal tensor diffusion coefficient and $\phi$
+is the diffusing variable. We use a central difference scheme to discretize this
+term on the face with inverse interpolation of each component of the diffusion tensor $k_i$:
+
+!equation
+1 / (k_i)_f = g / (k_i)_C + (1 - g) / (k_i)_N,
```
[the remark this answers is not in the document]
Comment (snschune) [reply framework/doc/content/source/fvkernels/FVAnisotropicDiffusion.md:15]: 1. Moukalled gives a good justification for it. We can discuss in tomorrow's meeting if you want.
2. It's exactly how I remember learning how to compute the average diffusion coefficient at an interface.
3. If you have heat conduction through two elements it's two resistors in series. The total resistance is the sum of both resistances and the resistances are inversely proportional to the thermal conductivities. So using the inverses makes physical sense as well. 

We can discuss tomorrow.

**Review by grmnptr — APPROVED** [review]

**Review by grmnptr — CHANGES_REQUESTED** [review]
I think I'll need to add changes here. I'll work on it later today. I'll try to separate the harmonic interpolation into another PR.

**Comment by grmnptr** [comment]
I'll finish this PR once we get a libmesh update and the general harmonic interpolation function is in!

File: framework/include/fvkernels/FVAnisotropicDiffusion.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "FVFluxKernel.h"
+
+/**
+ * FVAnisotropicDiffusion implements a standard diffusion term
+ * but with a diagonal tensor diffusion coefficient (provided as vector)
```
Comment (lindsayad) [inline framework/include/fvkernels/FVAnisotropicDiffusion.h:16]:
```suggestion
 * but with a diagonal tensor diffusion coefficient (provided as a vector)
```

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                "All relationship manager parameters should be valid.");
+
+    auto rm_obj = factory.create<RelationshipManager>(
+        "ElementSideNeighborLayers", name() + "_skew_correction", rm_params);
+
+    // Delete the resources created on behalf of the RM if it ends up not being added to the
+    // App.
+    if (!_app.addRelationshipManager(rm_obj))
+      factory.releaseSharedObjects(*rm_obj);
+  }
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:73]: We have this in 2-3 places now (maybe more). Might be good to to put this in the base `FVFluxKernel` class at some point

File: framework/src/fvkernels/FVAnisotropicDiffusion.C [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+ADReal
+FVAnisotropicDiffusion::computeQpResidual()
+{
+  const auto & grad_T = _var.adGradSln(
+      *_face_info, _var.faceInterpolationMethod() == Moose::FV::InterpMethod::SkewCorrectedAverage);
+
+  ADRealVectorValue k;
```
Comment (lindsayad) [inline framework/src/fvkernels/FVAnisotropicDiffusion.C:82]: Maybe just call this `coeff`

Comment (lindsayad) [reply framework/src/fvkernels/FVAnisotropicDiffusion.C:82]: I know this is the same naming convection as in `FVDiffusion`, but I think it's a little silly there as well

File: test/tests/fvkernels/fv_anisotropic_diffusion/tests [hunk 11]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  [fv_anisotropic_diffusion]
+    type = 'Exodiff'
+    input = 'fv_anisotropic_diffusion.i'
+    exodiff = 'fv_anisotropic_diffusion_out.e'
+    requirement = 'The system shall be able to solve anisotropic diffusion problems using the FV method'
```
Comment (lindsayad) [inline test/tests/fvkernels/fv_anisotropic_diffusion/tests:6]:
```suggestion
    requirement = 'The system shall be able to solve anisotropic diffusion problems using the finite volume method.'
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Review by grmnptr — DISMISSED** [review]

File: framework/src/fvkernels/FVAnisotropicDiffusion.C (hunk 9 above)
Comment (grmnptr) [reply framework/src/fvkernels/FVAnisotropicDiffusion.C]: Moved it to FVFluxKernel!

**Review by lindsayad — CHANGES_REQUESTED** [review]
This is awesome. Last request: can you use the new `adjustRMGhostLayers` interfaces in `INSFVAdvectionKernel` and `INSFVMomentumDiffusion`?

**Review by lindsayad — APPROVED** [review]

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**